File: cpp/leetcode/085.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if(matrix.empty() || matrix[0].empty()) return 0;
        int m = matrix.size(), n = matrix[0].size(), res = 0;
        vector<int> dp(n, 0);
        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                dp[j] = matrix[i][j] == '1' ? dp[j] + 1 : 0;
            }
            res = max(res, dfs(dp));
        }
        return res;
    }
    
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        int res = 0;
        stack<int> stk;
        for(int i = 0; i < heights.size(); i++){
            if(stk.empty()) stk.push(i);
            while(!stk.empty() && heights[i] < heights[stk.top()]){
                int idx = stk.top(); stk.pop();
                int width = stk.empty() ? i : (i - stk.top() - 1);
                res = max(heights[idx] * width, res);
            }
            stk.push(i);
        }
        return res;
    }
};

File: python/leetcode/085.py
class Solution:
    def largestRectangleArea(self, height):
        stack, i, res = [], 0, 0
        h = height + [0]
        while i < len(h):
            if not stack or h[stack[-1]] <= h[i]:
                stack.append(i)
                i += 1
            else:
                t = stack.pop()
                res = max(res, h[t] * (i if not stack else i - stack[-1] - 1))
        return res

    def maximalRectangle(self, matrix):
        if not matrix: return 0
        dp = [0 for i in range(len(matrix[0]))]
        maxArea = 0
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                dp[j] = dp[j] + 1 if matrix[i][j] == '1' else 0
            maxArea = max(maxArea, self.largestRectangleArea(dp))
        return maxArea